internal/uploader/events_uploader.go

```diff
@@ -0,0 +1,261 @@
+	buffer := &bytes.Buffer{}
+	gz := gzip.NewWriter(buffer)
+	tw := tar.NewWriter(gz)
+	failedFilesCount := 0
+
+	// errors creating files below will be ignored since failing to create one of the files
+	// doesn't mean the rest of the function should fail
+	cluster, err := clusterFile(db, tw, clusterID, pullSecret)
+	if err != nil {
+		log.WithError(err).Warnf("failed finding cluster events with cluster ID %s", clusterID)
+		failedFilesCount++
+	}
+
+	infraEnvID, err := hostsFile(db, tw, cluster)
+	if err != nil {
+		log.WithError(err).Warnf("failed finding host(s) events associated with cluster ID %s", clusterID)
+		failedFilesCount++
+	}
+
+	if err := infraEnvFile(db, tw, infraEnvID, clusterID); err != nil {
+		log.WithError(err).Warnf("failed finding infra-env events with ID %s", infraEnvID)
+		failedFilesCount++
+	}
+
+	if err := eventsFile(ctx, clusterID, eventsHandler, tw); err != nil {
+		log.WithError(err).Warnf("failed finding events associated with cluster ID %s", clusterID)
+		failedFilesCount++
+	}
+
+	// There are no files to upload
+	if failedFilesCount >= 4 {
+		return nil, errors.Errorf("no event data files created for cluster %s", clusterID)
+	}
+
+	// produce tar
+	if err := tw.Close(); err != nil {
+		return nil, errors.Wrapf(err, "failed closing tar file")
+	}
+	// produce gzip
+	if err := gz.Close(); err != nil {
+		return nil, errors.Wrapf(err, "failed closing gzip file")
+	}
+	return buffer, nil
+}
+
+func eventsFile(ctx context.Context, clusterID *strfmt.UUID, eventsHandler eventsapi.Handler, tw *tar.Writer) error {
+	if eventsHandler == nil {
+		return errors.Errorf("failed to get events for cluster %s, events handler is nil", clusterID)
+	}
+	dbEvents, err := eventsHandler.V2GetEvents(ctx, clusterID, nil, nil, models.EventCategoryMetrics, models.EventCategoryUser)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find events for cluster %s", *clusterID)
+	}
+
+	var events []*models.Event
+	for _, dbEvent := range dbEvents {
+		events = append(events, &dbEvent.Event)
+	}
+
+	contents, err := json.MarshalIndent(events, "", " ")
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal events")
+	}
+	return addFile(tw, contents, fmt.Sprintf("%s/events.json", clusterID))
+}
+
+func clusterFile(db *gorm.DB, tw *tar.Writer, clusterID *strfmt.UUID, pullSecret *serializedAuth) (*common.Cluster, error) {
+	if clusterID != nil {
+		cluster, err := common.GetClusterFromDBWithHosts(db, *clusterID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get cluster %s from db", *clusterID)
+		}
+		// To distinguish who is uploading the data
+		if cluster.EmailDomain == "" {
+			cluster.EmailDomain = getEmailDomain(pullSecret.Email)
+		}
+		if cluster.UserName == "" {
+			if token, tokenErr := ensureTokenFormat(pullSecret.Auth); tokenErr == nil {
+				cluster.UserName = getUsernameFromToken(token)
+			}
```

Comment:
do we need some warning log in case of tokenErr?

Reply:
I actually ended up removing this since the `ParsePullSecret` function you mentioned gives us the username directly!